keps/sig-api-machinery/4595-cel-crd-additionalprintercolumns/README.md

```diff
@@ -0,0 +1,1447 @@
+  }
+  path.AllowMissingKeys(true)
+  c.additionalColumns = append(c.additionalColumns, path)
+}
+```
+
+We then call this function in `TableConvertor.New()` to allow handling additionalPrinterColumns defined using CEL expressions:
+
+```diff
++ func New(crdColumns []apiextensionsv1.CustomResourceColumnDefinition, s *schema.Structural) (rest.TableConvertor, error) {
+  ...
++   if len(col.JSONPath) > 0 && len(col.Expression) == 0 {
+      // existing jsonPath logic
++   } else if len(col.Expression) > 0 && len(col.JSONPath) == 0 {
++     compResult := CompileColumn(col.Expression, s, model.SchemaDeclType(s, true), celconfig.PerCallLimit, environment.MustBaseEnvSet(environment.DefaultCompatibilityVersion(), true), cel.StoredExpressionsEnvLoader())
+
++     if compResult.Error != nil {
++       return c, fmt.Errorf("CEL compilation error %q", compResult.Error)
++      }
++     c.additionalColumns = append(c.additionalColumns, compResult)
++   }
+}
+```
+
+To make all this work, we also introduce the following:
+
+- A new struct `ColumnCompilationResult`:
+
+```go
+type ColumnCompilationResult struct {
+  Error          error
+  MaxCost        uint64
+  MaxCardinality uint64
+  FieldPath      *field.Path
+  Program        cel.Program
+}
+```
+
+- This struct implements the [columnPrinter](https://github.com/kubernetes/kubernetes/blob/8a0f6370e4b53b648050c534f0ee11b776f900a6/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/tableconvertor/tableconvertor.go#L74-L77) interface:
+
+```go
+func (c ColumnCompilationResult) FindResults(data interface{}) ([][]reflect.Value, error) {
+  ...
+}
+
+func (c ColumnCompilationResult) PrintResults(w io.Writer, results []reflect.Value) error {
+  ...
+}
+```
+
+- The output of `cel.CompileColumn()` returns a `ColumnCompilationResult` object for each additionalPrinterColumn. 
+
+With all of this we can pass the CEL program to the TableConvertor's `ConvertToTable()` method, which will call `FindResults` and `PrintResults` for all additionalPrinterColumns, regardless of whether they're defined with JSONPath or CEL expressions.
+
+
+### CEL vs JSONPath Performance Analysis
+
+A big part of the discussions for our proposal was the CEL cost limits since this is the first time CEL is added to the read path. As part of this we've done some benchmarking of the time it takes to parse and compile equivalent JSONPath and CEL expressions.
+
+For the performance analysis, we introduced two new benchmark tests to the `tableconvertor_test.go` file – `Benchmark_CEL` and `Benchmark_JSONPath`.
+
+Please find the raw output of the benchmark tests, as well as the code for executing them in the following gist:
+[https://gist.github.com/sreeram-venkitesh/f4aff1ae7957a5a3b9c6c53e869b7403](https://gist.github.com/sreeram-venkitesh/f4aff1ae7957a5a3b9c6c53e869b7403)
+
+
+Considering the following 10 iterations of the `Benchmark_CEL` and `Benchmark_JSONPath` tests:
+
+```
+  CEL Benchmark Results (10 runs):
```

Comment:
We should be clear that this also includes *compiling* CEL, which shouldn't happen in the hot loop in production.

I see the data below that splits these out, we should be clear about that.